### PR TITLE
Compute true loss Flax examples

### PR DESCRIPTION
**Pull request by duongna21**
# What does this PR do?

'True' losses should be computed in Flax examples, as [discussed](https://github.com/huggingface/transformers/pull/18297#discussion_r931971230) with @sanchit-gandhi.

## Who can review?
cc @sanchit-gandhi @patrickvonplaten

### CODE REVIEW

File: examples/flax/language-modeling/run_bart_dlm_flax.py [hunk 1]
```diff
@@ -888,7 +894,7 @@ def eval_step(params, batch):
                 num_eval_samples = len(tokenized_datasets["validation"])
                 # Avoid using jax.numpy here in case of TPU training
                 eval_samples_idx = np.arange(num_eval_samples)
-                eval_batch_idx = generate_batch_splits(eval_samples_idx, eval_batch_size, drop_last=False)
+                eval_batch_idx = generate_batch_splits(eval_samples_idx, eval_batch_size)
```
Comment (sanchit-gandhi) [inline examples/flax/language-modeling/run_bart_dlm_flax.py:897]: Unintentional change?

File: examples/flax/language-modeling/run_bart_dlm_flax.py [hunk 2]
```diff
@@ -928,7 +934,7 @@ def eval_step(params, batch):
         num_eval_samples = len(tokenized_datasets["validation"])
         # Avoid using jax.numpy here in case of TPU training
         eval_samples_idx = np.arange(num_eval_samples)
-        eval_batch_idx = generate_batch_splits(eval_samples_idx, eval_batch_size, drop_last=False)
+        eval_batch_idx = generate_batch_splits(eval_samples_idx, eval_batch_size)
```
Comment (sanchit-gandhi) [inline examples/flax/language-modeling/run_bart_dlm_flax.py:937]: Unintentional change?

File: examples/flax/language-modeling/run_bart_dlm_flax.py [hunk 3]
```diff
@@ -799,19 +799,25 @@ def loss_fn(params):
...
+        grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+        (loss, num_labels), grad = grad_fn(state.params)
+        num_labels = jax.lax.psum(num_labels, "batch")
 
-        metrics = jax.lax.pmean(
-            {"loss": loss, "learning_rate": linear_decay_lr_schedule_fn(state.step)}, axis_name="batch"
-        )
+        # true loss = total loss / total samples
+        loss = jax.lax.psum(loss, "batch")
+        loss = jax.tree_map(lambda x: x / num_labels, loss)
```
Comment (sanchit-gandhi) [inline examples/flax/language-modeling/run_bart_dlm_flax.py:813]: (nit) `jax.tree_map` is deprecated and will be replaced by `jax.tree_util.tree_map` in future versions of JAX
```suggestion
        loss = jax.tree_util.tree_map(lambda x: x / num_labels, loss)
```

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @duongna21, thanks for the PR - the 'true loss' implementation looks clean! There appear to be a few unintentional changes that have crept into the PR (deletion of lines, changes to `drop_last_batch`. It might simply be a case of rebasing onto main to fix these:
```
git fetch upstream
git rebase upstream/main
```

File: examples/flax/language-modeling/run_bart_dlm_flax.py (hunk 1 above)
Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: @sanchit-gandhi I noticed that with `drop_last=False` the eval loss will become `nan` at the beginning of training, but eval accuracy is still on track. It appears to occur with both `run_bart_dlm_flax` and `run_summarization` so I temporarily turned it off. It would be great if you could take a look and fix it.

File: examples/flax/language-modeling/run_bart_dlm_flax.py (hunk 3 above)
Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py]: Thanks! I've replaced all.

File: examples/flax/language-modeling/run_bart_dlm_flax.py (hunk 1 above)
Comment (sanchit-gandhi) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: Interesting! What immediately jumps out to me is that `num_labels` is 0, causing the 'true loss' to be `nan`. You didn't get this behaviour previously with the `pmap` operation? What eval batch size are you using?

Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: This bug occurs without dividing `loss` by `num_labels` actually. I took a quick look at the eval loss computed on every token at each step, the last losses was `nan`. 

```bash
python run_summarization_flax.py \
	--output_dir ./bart-base-xsum \
	--model_name_or_path facebook/bart-base \
	--tokenizer_name facebook/bart-base \
	--dataset_name="xsum" \
	--do_train --do_eval --do_predict --predict_with_generate \
	--num_train_epochs 6 \
	--learning_rate 5e-5 --warmup_steps 0 \
	--per_device_train_batch_size 64 \
	--per_device_eval_batch_size 64 \
	--overwrite_output_dir \
	--max_source_length 512 --max_target_length 64 \
	--push_to_hub 
```
Printed tensor has a shape of (8, 64, 64), provided that I have 8 TPU cores, `per_device_eval_batch_size=64` and `max_target_length=64`.
ảnh

Comment (sanchit-gandhi) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: Oh no! That's interesting to see. Is this an artefact of using the `psum`? As in, were the losses `nan` when we used a `pmap` previously? If so, we'll need to address this!

Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: The losses were `nan` right after `loss = optax.softmax_cross_entropy(logits, soft_labels)`. Great if you could have a look at this issue!

Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: Anyway, this problem doesn't seem to be related to this PR. Flax examples are young so we'll improve it step by step :)

Comment (tannonk) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: Hi @duongna21, just wondering what the status is on the `nan` issue being discussed here. I'm running into this issue while using the recently added `run_bart_dlm_flax.py` script but am very new to Flax/Jax so haven't been able to really make sense of it yet. Is `run_bart_dlm_flax.py` useable for model pre-training in its current state? Just as a side note, I haven't seen the `nan` issue when running `run_t5_mlm_flax.py` on my training data. Thanks in advance for any clarification!

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @duongna21! Sorry for the late review here. In general this looks good to me! Two things from my review:
1. Could you rebase onto main? This should update all the `jax.tree_map` calls to `jax.tree_util.tree_map`
2. Is using `drop_last=False` giving `nan` for the eval loss? If so, we can dig into this on a separate issue. Ideally we should keep the changes for this PR related to the 'true' Flax loss. We can open a separate issue/PR for any other side issues!

File: examples/flax/language-modeling/run_bart_dlm_flax.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:897]: @sanchit-gandhi Thanks for the review. I rebased.

File: examples/flax/language-modeling/run_bart_dlm_flax.py (hunk 2 above)
Comment (duongna21) [reply examples/flax/language-modeling/run_bart_dlm_flax.py:937]: @sanchit-gandhi Yes!

**Comment by patrickvonplaten** [comment]
@duongna21 super sorry it seems like the git commit history got messed up :-/ Any chance you could re-submit your PR?
